### DISCUSSION TITLE
Puppet manager with git@ git-tags can't fetch changelog with customChangelogUrl

### DISCUSSION THREAD

**Discussion by MindTooth** [renovatebot/renovate#21925] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.67.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

When setting `customChangelogUrl` in `packageRules` for the Puppet manager using git-tags, and `git@` addresses, a changelog is not fetched.

### What you tried so far.

I tried tweaking the regex, rules, etc.  But no go.  Viewing from the logs, `sourceUrl` is missing and the docs says the following:

> Source URLs are necessary in order to look up release notes.

https://docs.renovatebot.com/configuration-options/#customchangelogurl

---

## Test Project

I replicated it using public GitHub and GitLab.

Project: https://github.com/MindTooth/renovate-git-tags-test

`http` address (working): https://github.com/MindTooth/renovate-git-tags-test/pull/7

`git@` address (not working): https://github.com/MindTooth/renovate-git-tags-test/pull/2

---

Please let me know if something is missing.  Been at this for months trying to figure this out.  Last resort was to open this discussion.

From viewing the source code, I've made the loose assumption that because a `sourceUrl` of the format `git@` isn't parseable, it is set to `nil` and when creating/looking up the changelog, it sees that the package does not have a `sourceUrl` and then skips it all together.  I found a `massageUrl` function that perhaps fixes this?

### Relevant debug logs

<details><summary>Logs</summary>

```json
          {
            "currentValue": "v6.0.0",
            "currentVersion": "v6.0.0",
            "datasource": "git-tags",
            "depName": "apache",
            "fixedVersion": "v6.0.0",
            "gitRef": true,
            "isSingleVersion": true,
            "packageName": "https://gitlab.com/hitchhikers/puppet-modules/puppetlabs-apache.git",
            "sourceUrl": "https://gitlab.com/hitchhikers/puppet-modules/puppetlabs-apache",
            "versioning": "semver",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "v6.5.1",
                "newValue": "v6.5.1",
                "newDigest": "3fff7d86c44f4e380172bd4317c50d60d0cba654",
                "newMajor": 6,
                "newMinor": 5,
                "updateType": "minor",
                "branchName": "renovate/apache-6.x"
              },
              {
                "bucket": "major",
                "newVersion": "v9.1.2",
                "newValue": "v9.1.2",
                "newDigest": "0f7fe11000b9d56c862070ec21b81245b1de326b",
                "newMajor": 9,
                "newMinor": 1,
                "updateType": "major",
                "branchName": "renovate/apache-9.x"
              }
            ]
          }
```
vs.
```json
          {
            "currentValue": "v6.0.0",
            "currentVersion": "v6.0.0",
            "datasource": "git-tags",
            "depName": "logrotate",
            "fixedVersion": "v6.0.0",
            "gitRef": true,
            "isSingleVersion": true,
            "packageName": "git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git",
            "versioning": "semver",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "v6.1.0",
                "newValue": "v6.1.0",
                "newDigest": "486384c19a1c1e9b1e4076218bde01806e7e058f",
                "newMajor": 6,
                "newMinor": 1,
                "updateType": "minor",
                "branchName": "renovate/logrotate-6.x"
              }
            ]
          }
```

https://app.renovatebot.com/dashboard#github/MindTooth/renovate-git-tags-test/1062412487

</details>

**Comment by rarkins** [comment]
Renovate only attempts to fetch and parse changelogs from known https hosts like github.com

**Reply by MindTooth** [reply]
So, setting a valid `customChangelogUrl` to GitHub does not work on its own?

```json
    {
      "matchManagers": ["puppet"],
      "matchPackagePatterns": [
        "gitlab\\.com[:/]hitchhikers/puppet-modules/voxpupuli-logrotate(?:\\.git)?$"
      ],
      "customChangelogUrl": "https://github.com/voxpupuli/puppet-logrotate"
    }
```

---

Before me attempting any contribution to the project.  Would you accept a fix if I manage to come up with one?  Because, I feel that when I tell renovate to use a changelog, I would expect it to work.  But, I might be missing the bigger picture here.

**Reply by rarkins** [reply]
I would expect setting a customer change log URL like that (i.e. to a github.com repo URL) to work

**Reply by MindTooth** [reply]
I'll look more into this then, and see if I'm able to contribute a fix if possible.  Some places I will start to look:

- See why the `PackageDependency` does not contain a `sourceUrl:` field.  And if this is the problem later down the chain.
- Let the manager set `sourceUrl` using `massageUrl()` to make it a HTTP variant, and then see if that solves anything.
- Look at `customChangelogUrl` and when it function/triggers in relation with `sourceUrl`.  Ref. the quote, it suggests that it needs a `sourceUrl` field.
- Document any findings to perhaps being more clear.  Initial understating is that not all managers/datasources provide a `sourceUrl` and that's why it can be missing.

---

Thanks for the replies.  Love the tool and a big promoter.  Just have this one issue that I've been struggling for a long time.

**Reply by MindTooth** [reply]
Now it's been over a month, and I've spent so much time trying to find this out. 🥵   I just now tried something for fun and it seems to actually show the "bug".  I'll try to be as specific as possible.

---

Given the following `Puppetfile` and the [Puppet manager](https://docs.renovatebot.com/modules/manager/puppet/#supported-puppetfile-formats):

```Puppefile
mod 'logrotate',
  :git => 'git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git',
  :tag => 'v3.3.0'
```

It will output the following:

```json
       "packageFiles": {
         "puppet": [
           {
             "deps": [
               {
                 "depName": "logrotate",
                 "packageName": "git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git",
                 "sourceUrl": "git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git",
                 "gitRef": true,
                 "currentValue": "v3.3.0",
                 "datasource": "git-tags"
               }
             ],
             "packageFile": "Puppetfile"
           }
         ]
       }
```

I have for months, and I really mean months, trying to understand why the following would not work when `packageName` clearly states the full URL as its `packageName`:
```json
    {
      "matchManagers": ["puppet"],
      "matchPackagePatterns": [
        "git@gitlab\\.com:hitchhikers/puppet-modules/voxpupuli-logrotate\\.git"
      ],
      "customChangelogUrl": "https://github.com/puppetlabs/puppetlabs-apache"
    },
```

_(The changelog URL for apache is acting as a canary for the rule.)_

What I tried was to use the whole URL as the module's `depName` and then the regex worked perfectly. 😭 

```Puppefile
mod 'git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git',
  :git => 'git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git',
  :tag => 'v3.3.0'
```

---

What I can gather is the following:

1.
```json
      "matchPackagePatterns": [
        "git@gitlab\\.com:hitchhikers/puppet-modules/voxpupuli-logrotate\\.git"
      ],
 ```
 is not equal to (`packageName` != `matchPackagePatterns`)
 ```json
 "packageName": "git@gitlab.com:hitchhikers/puppet-modules/voxpupuli-logrotate.git"
 ```
2. `matchPackagePatterns` works against `depName` and not `packageName`
3. Some other bugs.

I really don't know anymore.  If this has been the issue all along, it makes me kinda sad. 😅  What am I doing wrong?  I can't for the life of my understand why it can't work when they both are named _package_ something.

---

* https://github.com/MindTooth/renovate-puppetfile-test
* https://github.com/MindTooth/renovate-git-tags-test

**Reply by MindTooth** [reply]
Is it simple down to `RENOVATE_X_MATCH_PACKAGE_NAMES_MORE` and what I'm trying to do does not work because `packageName` isn't included in the matching?

https://github.com/renovatebot/renovate/blob/0c18515d3806b7dbab0bb75ac806565e2ef82962/lib/util/package-rules/package-patterns.ts#L23-L28

---

However, does not explain why `sourceUrl` isn't shown in the output when using the `git-tags` datasource.

**Reply by rarkins** [reply, marked as answer]
Each dependency should have a depName and packageName. `matchDepNames` can be used to match against `depName` while `matchPackageNames` can be used to match against `packageName` too if you have `RENOVATE_X_MATCH_PACKAGE_NAMES_MORE` set.

**Reply by MindTooth** [reply]
Is this to save on processing needs?  To not match on both I mean.

**Reply by rarkins** [reply]
No. It's for legacy/ backwards compatibility reasons and will be changed in the next major release

**Reply by MindTooth** [reply]
So, to close off, the next major release will fix this and I won't need to export the variable anymore (meaning the same behavior will work with the hosted version here on GitHub)?

A major thanks for your replies. 😁

**Reply by rarkins** [reply]
I can't confirm what it will fix, but in the next major release matchPackageNames will match against both packageName and depName by default

**Reply by MindTooth** [reply]
I guess that also works for `matchPackagePatterns`.

Thanks again.  Looking forward to the new release soon. 😄
